File: winanalytics-retrofit/src/main/java/com/winfooz/AnalyticsFactory.java
package com.winfooz;

import android.support.annotation.NonNull;
import java.lang.annotation.Annotation;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import retrofit2.Call;
import retrofit2.CallAdapter;
import retrofit2.Retrofit;

public class AnalyticsFactory extends CallAdapter.Factory {

    private final String baseUrl;

    public AnalyticsFactory(String baseUrl) {
        this.baseUrl = baseUrl;
    }

    @Override
    public CallAdapter<?, ?> get(
            @NonNull Type returnType,
            @NonNull Annotation[] annotations,
            @NonNull Retrofit retrofit) {
        if (getRawType(returnType) != Call.class) {
            return null;
        }
        if (!(returnType instanceof ParameterizedType)) {
            String msg = "Call must have generic type (e.g., Call<ResponseBody>)";
            throw new IllegalStateException(msg);
        }
        Type responseType = getParameterUpperBound(0, (ParameterizedType) returnType);
        boolean logEvent = false;
        for (Annotation annotation : annotations) {
            if (annotation.annotationType().equals(AnalyticsCall.class)) {
                logEvent = true;
                break;
            }
        }
        return new AnalyticsCallAdapter<>(responseType, baseUrl, logEvent);
    }
}

File: winanalytics-retrofit/src/main/java/com/winfooz/AnalyticsFactory.kt
package com.winfooz

import retrofit2.Call
import retrofit2.CallAdapter
import retrofit2.Retrofit
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

class AnalyticsFactory(private val baseUrl: String) : CallAdapter.Factory() {

    override fun get(
        returnType: Type,
        annotations: Array<Annotation>,
        retrofit: Retrofit
    ): CallAdapter<*, *>? {
        if (CallAdapter.Factory.getRawType(returnType) != Call::class.java) {
            return null
        }
        if (returnType !is ParameterizedType) {
            val msg = "Call must have generic type (e.g., Call<ResponseBody>)"
            throw IllegalStateException(msg)
        }
        val responseType = CallAdapter.Factory.getParameterUpperBound(0, returnType)
        return AnalyticsCallAdapter<Any>(
            responseType,
            baseUrl,
            annotations.any { it.annotationClass == AnalyticsCall::class }
        )
    }
}
